pam/gdm/protocol_test.go

```diff
@@ -0,0 +1,672 @@
+		},
+
+		"Error response packet with missing type": {
+			gdmData: &Data{
+				Type:     DataType_response,
+				Response: &ResponseData{Data: &ResponseData_Ack{}},
+			},
+
+			wantError: "missing response type",
+		},
+
+		"Error response packet with invalid type": {
+			gdmData: &Data{
+				Type:     DataType_response,
+				Response: &ResponseData{Type: RequestType(-1), Data: &ResponseData_Ack{}},
+			},
+
+			wantError: "unexpected request type -1",
+		},
+
+		"Error response packet with unexpected data": {
+			gdmData: &Data{
+				Type:     DataType_response,
+				Response: &ResponseData{Type: RequestType_changeStage, Data: &ResponseData_Ack{}},
+				Event:    &EventData{},
+			},
+
+			wantError: "field Event should not be defined",
+		},
+
+		"Error poll packet with unexpected data": {
+			gdmData: &Data{Type: DataType_poll, Request: &RequestData{}},
+
+			wantError: "field Request should not be defined",
+		},
+
+		"Error pollResponse packet with missing event type": {
+			gdmData: &Data{
+				Type: DataType_pollResponse,
+				PollResponse: []*EventData{
+					{
+						Type: EventType_brokerSelected,
+						Data: &EventData_BrokerSelected{
+							BrokerSelected: &Events_BrokerSelected{BrokerId: "a broker"},
+						},
+					},
+					{
+						Data: &EventData_AuthModeSelected{
+							AuthModeSelected: &Events_AuthModeSelected{AuthModeId: "auth mode"},
+						},
+					},
+				},
+			},
+
+			wantError: "poll response data member 1 invalid: missing event type",
+		},
+
+		"Error pollResponse packet with event with missing type": {
+			gdmData: &Data{
+				Type: DataType_pollResponse,
+				PollResponse: []*EventData{
+					{},
+					{
+						Type: EventType_authModeSelected,
+						Data: &EventData_AuthModeSelected{
+							AuthModeSelected: &Events_AuthModeSelected{AuthModeId: "auth mode"},
+						},
+					},
+				},
+			},
+
+			wantError: "poll response data member 0 invalid: missing event type",
+		},
+
+		"Error pollResponse packet with unexpected data": {
+			gdmData: &Data{
+				Type:         DataType_pollResponse,
+				PollResponse: []*EventData{},
+				Event:        &EventData{},
+			},
+
+			wantError: "field Event should not be defined",
+		},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			bytes, err := tc.gdmData.JSON()
+			if tc.wantError != "" {
+				require.ErrorContains(t, err, tc.wantError)
+				return
+			}
+
+			require.NoError(t, err)
+			formattedJSON := string(reformatJSON(t, bytes))
+			require.Equal(t, tc.wantJSON, formattedJSON)
+
+			// Now try to reconvert things back again
+			gdmData, err := NewDataFromJSON(bytes)
+			require.NoError(t, err)
+			newBytes, err := gdmData.JSON()
+			require.NoError(t, err)
+			require.Equal(t, formattedJSON, string(reformatJSON(t, newBytes)))
+		})
+	}
+}
+
+func TestGdmStructsUnMarshal(t *testing.T) {
+	t.Parallel()
+	tests := map[string]struct {
+		JSON               string
+		ignoreReconversion bool
+
```

Comment:
nitpick: `ignoreReconversion` is a "test flow controller" rather than a function input, so I think it should be on its own "block", e.g:
```Go
JSON string

ignoreReconversion bool

wantData *Data
wantError string 
```

Reply:
oh IIRC this can be avoided at all now. I had included it when I [was defining the protocol myself by hand](https://github.com/3v1n0/authd/blob/gdm-json-protocol/pam/gdm/protocol.go), where native `json` may not be reliable on reconverting back at times.